File: api/app/schemas/disponibilidad.py
from pydantic import BaseModel, ConfigDict, field_validator
from uuid import UUID
from datetime import date, datetime
from decimal import Decimal

ESTADOS_DISPONIBILIDAD = ["pendiente", "aprobado", "rechazado"]


class DisponibilidadBase(BaseModel):
    entidad_id: UUID
    centro_costo_id: UUID
    poi_id: UUID
    meta_id: UUID | None = None
    clasificacion_id: UUID
    fuente_id: UUID
    anio_fiscal: int
    numero_solicitud: str
    monto_solicitado: Decimal
    descripcion: str | None = None
    fecha_solicitud: date

    @field_validator("numero_solicitud")
    @classmethod
    def validar_numero(cls, v: str) -> str:
        v = v.strip()
        if len(v) < 3:
            raise ValueError("El número de solicitud es muy corto")
        return v

    @field_validator("monto_solicitado")
    @classmethod
    def validar_monto(cls, v: Decimal) -> Decimal:
        if v <= 0:
            raise ValueError("El monto debe ser mayor a 0")
        return v

    @field_validator("anio_fiscal")
    @classmethod
    def validar_anio(cls, v: int) -> int:
        if v < 2000 or v > 2100:
            raise ValueError("Año fiscal fuera de rango (2000-2100)")
        return v


class DisponibilidadCreate(DisponibilidadBase):
    pass


class DisponibilidadUpdate(BaseModel):
    numero_solicitud: str | None = None
    monto_solicitado: Decimal | None = None
    descripcion: str | None = None
    meta_id: UUID | None = None


class DisponibilidadAprobar(BaseModel):
    monto_aprobado: Decimal
    observaciones: str | None = None

    @field_validator("monto_aprobado")
    @classmethod
    def validar_monto_aprobado(cls, v: Decimal) -> Decimal:
        if v < 0:
            raise ValueError("El monto aprobado no puede ser negativo")
        return v


class DisponibilidadRechazar(BaseModel):
    observaciones: str

    @field_validator("observaciones")
    @classmethod
    def validar_observaciones(cls, v: str) -> str:
        if len(v.strip()) < 10:
            raise ValueError("Las observaciones deben tener al menos 10 caracteres")
        return v


class DisponibilidadOut(DisponibilidadBase):
    id: UUID
    monto_aprobado: Decimal
    estado: str
    observaciones: str | None = None
    aprobado_por: UUID | None = None
    fecha_aprobacion: datetime | None = None
    creado_en: datetime
    actualizado_en: datetime | None = None
    
    # Campos relacionados (JOINs)
    entidad_nombre: str | None = None
    centro_costo_nombre: str | None = None
    poi_nombre: str | None = None
    clasificador_codigo: str | None = None
    clasificador_descripcion: str | None = None
    fuente_nombre: str | None = None

    model_config = ConfigDict(from_attributes=True)

File: web/src/types/disponibilidad.ts
import { z } from 'zod';

export const ESTADOS_DISPONIBILIDAD = ['pendiente', 'aprobado', 'rechazado'] as const;
export type EstadoDisponibilidad = (typeof ESTADOS_DISPONIBILIDAD)[number];

export const disponibilidadSchema = z.object({
  entidad_id: z.string().uuid('Debe seleccionar una entidad'),
  centro_costo_id: z.string().uuid('Debe seleccionar un centro de costo'),
  poi_id: z.string().uuid('Debe seleccionar un POI'),
  meta_id: z.string().uuid().optional().or(z.literal('')),
  clasificacion_id: z.string().uuid('Debe seleccionar un clasificador'),
  fuente_id: z.string().uuid('Debe seleccionar una fuente'),
  anio_fiscal: z.coerce.number().min(2000).max(2100),
  numero_solicitud: z.string().min(3, 'Mínimo 3 caracteres').max(50),
  monto_solicitado: z.coerce.number().min(0.01, 'Debe ser mayor a 0'),
  descripcion: z.string().optional().or(z.literal('')),
  fecha_solicitud: z.string().min(1, 'La fecha es requerida'),
});

export type DisponibilidadFormValues = z.infer<typeof disponibilidadSchema>;

export interface DisponibilidadEntity {
  id: string;
  entidad_id: string;
  centro_costo_id: string;
  poi_id: string;
  meta_id?: string;
  clasificacion_id: string;
  fuente_id: string;
  anio_fiscal: number;
  numero_solicitud: string;
  monto_solicitado: number;
  monto_aprobado: number;
  descripcion?: string;
  observaciones?: string;
  estado: EstadoDisponibilidad;
  fecha_solicitud: string;
  aprobado_por?: string;
  fecha_aprobacion?: string;
  creado_en: string;
  actualizado_en?: string;
  // Campos relacionados
  entidad_nombre?: string;
  centro_costo_nombre?: string;
  poi_nombre?: string;
  clasificador_codigo?: string;
  clasificador_descripcion?: string;
  fuente_nombre?: string;
}

export interface DisponibilidadPaginatedResponse {
  data: DisponibilidadEntity[];
  total: number;
  page: number;
  limit: number;
  totalPages: number;
}

export interface DisponibilidadListParams {
  page: number;
  limit: number;
  search?: string;
  sort_by?: string;
  sort_order?: 'asc' | 'desc';
  entidad_id?: string;
  estado?: string;
  anio_fiscal?: number;
}